experiment/simulation/assets/DragControls.js

```diff
@@ -0,0 +1,233 @@
+
+        _domElement.style.cursor = "auto";
+        _hovered = null;
+      }
+    }
+  }
+
+  function onDocumentMouseDown(event) {
+    event.preventDefault();
+
+    _raycaster.setFromCamera(_mouse, _camera);
+
+    var intersects = _raycaster.intersectObjects(_objects);
+
+    if (intersects.length > 0) {
+      _selected = intersects[0].object;
+
+      if (_raycaster.ray.intersectPlane(_plane, _intersection)) {
+        _offset.copy(_intersection).sub(_selected.position);
+      }
+
+      _domElement.style.cursor = "move";
+
+      scope.dispatchEvent({ type: "dragstart", object: _selected });
+    }
+  }
+
+  function onDocumentMouseCancel(event) {
+    event.preventDefault();
+
+    if (_selected) {
+      scope.dispatchEvent({ type: "dragend", object: _selected });
+
+      _selected = null;
+    }
+
+    _domElement.style.cursor = "auto";
+  }
+
+  function onDocumentTouchMove(event) {
+    event.preventDefault();
+    event = event.changedTouches[0];
+
+    var rect = _domElement.getBoundingClientRect();
+
+    _mouse.x = ((event.clientX - rect.left) / rect.width) * 2 - 1;
+    _mouse.y = -((event.clientY - rect.top) / rect.height) * 2 + 1;
+
+    _raycaster.setFromCamera(_mouse, _camera);
+
+    if (_selected && scope.enabled) {
+      if (_raycaster.ray.intersectPlane(_plane, _intersection)) {
+        _selected.position.copy(_intersection.sub(_offset));
+      }
+
+      scope.dispatchEvent({ type: "drag", object: _selected });
+
+      return;
+    }
+  }
+
+  function onDocumentTouchStart(event) {
+    event.preventDefault();
+    event = event.changedTouches[0];
+
+    var rect = _domElement.getBoundingClientRect();
+
+    _mouse.x = ((event.clientX - rect.left) / rect.width) * 2 - 1;
+    _mouse.y = -((event.clientY - rect.top) / rect.height) * 2 + 1;
+
+    _raycaster.setFromCamera(_mouse, _camera);
+
+    var intersects = _raycaster.intersectObjects(_objects);
+
+    if (intersects.length > 0) {
+      _selected = intersects[0].object;
+
+      _plane.setFromNormalAndCoplanarPoint(
+        _camera.getWorldDirection(_plane.normal),
+        _selected.position
+      );
+
+      if (_raycaster.ray.intersectPlane(_plane, _intersection)) {
+        _offset.copy(_intersection).sub(_selected.position);
+      }
+
+      _domElement.style.cursor = "move";
+
+      scope.dispatchEvent({ type: "dragstart", object: _selected });
+    }
+  }
+
+  function onDocumentTouchEnd(event) {
+    event.preventDefault();
+
+    if (_selected) {
+      scope.dispatchEvent({ type: "dragend", object: _selected });
+
+      _selected = null;
+    }
+
+    _domElement.style.cursor = "auto";
+  }
+
+  activate();
+
+  // API
+
+  this.enabled = true;
+
+  this.activate = activate;
+  this.deactivate = deactivate;
+  this.dispose = dispose;
+
+  // Backward compatibility
+
+  this.setObjects = function () {
+    console.error("THREE.DragControls: setObjects() has been removed.");
+  };
+
+  this.on = function (type, listener) {
```

Comment:
Why do we have deprecated functions here? Has this code been copied from somewhere?